bobgit/commit.go

```diff
@@ -0,0 +1,125 @@
+package bobgit
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/Benchkram/bob/pkg/boblog"
+	"github.com/Benchkram/bob/pkg/bobutil"
+	"github.com/Benchkram/bob/pkg/cmdutil"
+	"github.com/Benchkram/bob/pkg/usererror"
+	"github.com/Benchkram/errz"
+	git "github.com/go-git/go-git/v5"
+)
+
+var ErrEmptyCommitMessage = fmt.Errorf("Bobgit does not allow empty message")
+
+// Commit executes `git commit -m ${message}` in all repositories.
+//
+// indifferent of the subdirectories and subrepositories,
+// it walks through all the repositories starting from bobroot
+// and run `git commit -m {message}` command.
+//
+// Only shows user messages in case of nothing to commit.
+func Commit(message string) (err error) {
+	defer errz.Recover(&err)
+
+	if message == "" {
+		return usererror.Wrapm(ErrEmptyCommitMessage, "Aborting commit due to empty commit message")
+	}
+
+	bobRoot, err := bobutil.FindBobRoot()
+	errz.Fatal(err)
+
+	err = os.Chdir(bobRoot)
+	errz.Fatal(err)
+
+	// Assure toplevel is a git repo
+	isGit, err := isGitRepo(bobRoot)
+	errz.Fatal(err)
+	if !isGit {
+		return usererror.Wrap(ErrCouldNotFindGitDir)
+	}
+
+	// search for git repos inside bobRoot/.
+	repoNames, err := findRepos(bobRoot)
+	errz.Fatal(err)
+
+	// repos with no changes, throws exit status 1 error
+	// while executing `git commit --dry-run`
+	// only repos with changes filtered out
+	filteredRepo, untrackedRepo, err := filterModifiedRepos(repoNames)
+	errz.Fatal(err)
+
+	// throw some user message for untracked repositories
+	if len(filteredRepo) == 0 {
+		s := "nothing to commit, working trees are clean."
+		if len(untrackedRepo) > 0 {
+			s = "nothing added to commit but untracked files present."
+		}
+		boblog.Log.V(1).Info(s)
```

Comment:
Do not use boblog for user facing messages, should only be used if you want to log something. Using `fmt.Printf()` for cli output is totaly fine.

Reply:
removed